src/target/lpc40xx.c

```diff
@@ -0,0 +1,273 @@
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "general.h"
+#include "target.h"
+#include "target_internal.h"
+#include "cortexm.h"
+#include "lpc_common.h"
+#include "adiv5.h"
+
+/*
+ * For detailed documentation on how this code works and the IAP variant used here, see:
+ * https://www.nxp.com/docs/en/data-sheet/LPC1759_58_56_54_52_51.pdf
+ * and (behind their login wall):
+ * https://cache.nxp.com/secured/assets/documents/en/user-guide/UM10360.pdf?fileExt=.pdf
+ */
+
+#define IAP_PGM_CHUNKSIZE 4096U
+
+#define MIN_RAM_SIZE               8192U // LPC1751
+#define RAM_USAGE_FOR_IAP_ROUTINES 32U   // IAP routines use 32 bytes at top of ram
+
+#define IAP_ENTRYPOINT 0x1fff1ff1U
+#define IAP_RAM_BASE   0x10000000U
+
+#define LPC40xx_MEMMAP   UINT32_C(0x400fc040)
+#define LPC40xx_MPU_BASE UINT32_C(0xe000ed90)
+#define LPC40xx_MPU_CTRL (LPC40xx_MPU_BASE + 0x04U)
+
+#define FLASH_NUM_SECTOR 30U
+
+typedef struct iap_config {
+	uint32_t command;
+	uint32_t params[4];
+} iap_config_s;
+
+typedef struct __attribute__((aligned(4))) iap_frame {
+	/* The start of an IAP stack frame is the opcode we set as the return point. */
+	uint16_t opcode;
+	/* There's then a hidden alignment field here, followed by the IAP call setup */
+	iap_config_s config;
+} iap_frame_s;
+
+typedef struct iap_result {
+	uint32_t return_code;
+	uint32_t values[4];
+} iap_result_s;
+
+typedef struct lpc40xx_priv {
+	uint32_t mpu_ctrl_state;
+	uint32_t memmap_state;
+} lpc40xx_priv_s;
+
+static void lpc40xx_extended_reset(target_s *target);
+static bool lpc40xx_enter_flash_mode(target_s *target);
+static bool lpc40xx_exit_flash_mode(target_s *target);
+static bool lpc40xx_mass_erase(target_s *target);
+iap_status_e lpc40xx_iap_call(target_s *target, iap_result_s *result, iap_cmd_e cmd, ...);
+
+static void lpc40xx_add_flash(target_s *target, uint32_t addr, size_t len, size_t erasesize, uint8_t base_sector)
+{
+	lpc_flash_s *flash = lpc_add_flash(target, addr, len, IAP_PGM_CHUNKSIZE);
+	flash->f.blocksize = erasesize;
+	flash->base_sector = base_sector;
+	flash->f.write = lpc_flash_write_magic_vect;
+	flash->iap_entry = IAP_ENTRYPOINT;
+	flash->iap_ram = IAP_RAM_BASE;
+	flash->iap_msp = IAP_RAM_BASE + MIN_RAM_SIZE - RAM_USAGE_FOR_IAP_ROUTINES;
+}
+
+bool lpc40xx_probe(target_s *target)
+{
+	if ((target->cpuid & CPUID_PARTNO_MASK) == CORTEX_M4) {
```

Comment:
Would this make more sense as a check for `!= CORTEX_M4` and a fast `return false;` so the rest of the code can be less indented?

Reply:
Fair point. I also fixed the lpc17xx.c that this was copied from.